### PR TITLE
[SPARK-21896][SQL] Fix StackOverflow caused by window functions inside aggregate functions

### PR DESCRIPTION
**Pull request by aokolnychyi**
## What changes were proposed in this pull request?

This PR explicitly prohibits window functions inside aggregates. Currently, this will cause StackOverflow during analysis. See PR #19193 for previous discussion.

## How was this patch tested?

This PR comes with a dedicated unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -1744,11 +1744,14 @@ class Analyzer(
...
   object ExtractWindowExpressions extends Rule[LogicalPlan] {
-    private def hasWindowFunction(projectList: Seq[NamedExpression]): Boolean =
-      projectList.exists(hasWindowFunction)
+    private def hasWindowFunction(exprs: Seq[Expression]): Boolean =
+      exprs.exists(hasWindowFunction)
 
-    private def hasWindowFunction(expr: NamedExpression): Boolean = {
+    private def hasWindowFunction(expr: Expression): Boolean = {
       expr.find {
+        case AggregateExpression(aggFunc, _, _, _) if hasWindowFunction(aggFunc.children) =>
```
Comment (aokolnychyi) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I have some doubts that this is the best place for this check. StackOverflow happens in ``extract``. We can also define a separate method and call it inside ``extract``. However, that method will share the same structure as ``hasWindowFunction``.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameAggregateSuite.scala [hunk 2]
```diff
@@ -687,4 +687,29 @@ class DataFrameAggregateSuite extends QueryTest with SharedSQLContext {
...
+    checkWindowError(testData2.groupBy('a).agg(max(sum(sum('b)).over(Window.orderBy('b)))))
+
+    checkWindowError(
+      sql("SELECT MAX(RANK() OVER(ORDER BY b)) FROM testData2 GROUP BY a HAVING SUM(b) = 3"))
+    checkWindowError(
+      sql("SELECT MAX(RANK() OVER(ORDER BY a)) FROM testData2"))
+    checkWindowError(
+      sql("SELECT MAX(RANK() OVER(ORDER BY b)) FROM testData2 GROUP BY a"))
+    checkAnswer(
+      sql("SELECT a, MAX(b), RANK() OVER(ORDER BY a) FROM testData2 GROUP BY a HAVING SUM(b) = 3"),
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameAggregateSuite.scala]: I think the dataset version should be 
```
df.groupBy('a).agg(max('b), sum('b).as("sumb"), rank().over(window)).where('sumb === 5)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: It's weird to throw exception inside a boolean function `hasXXX`.

Can we do this check in https://github.com/apache/spark/pull/21473/files#diff-57b3d87be744b7d79a9beacf8e5e5eb2R1838 ?

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by vincent-grosbois** [comment]
Hello!
Is there anyway this can be backported in 2.3?

**Pull request closed without merging**